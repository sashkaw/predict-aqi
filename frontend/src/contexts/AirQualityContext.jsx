import { createContext, useContext, useState, useEffect } from "react";
import axios from "axios";

export const AQIContext = createContext();
export const useAQIContext = () => useContext(AQIContext);

export const AQIProvider = (props) => {
    const [site, setSite] = useState({
        "id": 1,
        "name": "Amazon Park - Eugene, OR",
        "latitude": 44.026280,
        "longitude": -123.083715,
      });

    const [currentData, setCurrentData] = useState(null);
    const [forecastData, setForecastData] = useState([]);
    const [futureTimeSteps, setFutureTimeSteps] = useState([]);

    // Fetch AQI data from backend after DOM has updated
    useEffect(() => {
        axios('/aqi/')
            .then((response) => {
                //console.log(response.data.aqi_current.at(-1))
                // Get most recent AQI observation
                setCurrentData(response.data.aqi_current.at(-1));
                // Get forecasted AQI for future time steps
                setForecastData(response.data.aqi_forecast);
                // Get corresponding time steps for forecasted data
                setFutureTimeSteps(response.data.future_timesteps);
                
            });
    }, [site]); // Only run if value of site changes
    //console.log(futureTimeSteps, currentData)

    return (
        <AQIContext.Provider
            value={{
                site, 
                setSite,
                currentData,
                setCurrentData,
                forecastData,
                setForecastData,
                futureTimeSteps,
                setFutureTimeSteps,
            }}
            >
                {props.children}
            </AQIContext.Provider>
    );

};



