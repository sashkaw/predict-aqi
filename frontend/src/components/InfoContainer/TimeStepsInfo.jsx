import { useEffect, useState } from "react";
import { useAQIContext } from "../../contexts/AirQualityContext";
import { DayCard } from "../DayCard";

export function TimeStepsInfo() {
    const { forecastData, futureTimeSteps, setCurrentData , setCurrentTime} = useAQIContext();
    const [selectedCard, setSelectedCard] = useState(0);

    useEffect(() => {
        setCurrentData(forecastData[selectedCard]);
        setCurrentTime(futureTimeSteps[selectedCard]);
    }, [forecastData]); // Skip effect if forecast data already set

    return (
        <div className="timesteps-container">
            <ul className="timesteps-list">
                {forecastData.map((item, index) => {
                    if(index < 12) { //TODO: change this to fit number of timesteps dynamically
                        return (
                            <DayCard
                                className={index === selectedCard ? "active": ""}
                                onClick={() => {
                                    setSelectedCard(index);
                                    setCurrentData(item);
                                    setCurrentTime(futureTimeSteps[index]);
                                }}
                                key={index}
                                index={index}
                                item={item}
                            />
                        );
                    } 
                })}
                <div className="clear"></div>
            </ul>
        </div>
    );
}


/*const aqi_level = (aqi) => {
    const level = {};
    if(aqi >= 0 && aqi <= 50) {
        level = {
            "good: green",
        };
    }
    else if(aqi <= 100) {
        level = {
            "moderate: yellow",
        };
    }
    else if(aqi <= 150) {
        level = {
            "unhealthy for sensitive groups": "orange",
        }
    }
    else if(aqi <= 200) {
        level = {
            "unhealthy": "red",
        }
    }
    else if(aqi <= 300) {
        level = {
            "very unhealthy": "purple",
        }
    }
    else if(aqi <= 500) {
        level = {
            "hazardous": "maroon",
        }
    }

    return level;
}*/