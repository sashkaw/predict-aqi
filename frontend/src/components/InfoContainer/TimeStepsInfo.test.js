import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { useEffect, useState, useContext } from "react";
import { AQIContext } from "../../contexts/AirQualityContext";
import TestAQIProvider from "../Fixtures";
import { TimeStepsInfo } from "./TimeStepsInfo";

// Set up Location with mock data
const Fixture = () => {
    const Child = () => {
        return(
            <TimeStepsInfo />
        )
    };
    // Return Location wrapped within Provider
    return (
        <TestAQIProvider>
            <Child />
        </TestAQIProvider>
    );
}

// Helper function for snapshot testing
const matches = (children) => expect(
    renderer.create(children).toJSON()
).toMatchSnapshot();

describe("<TimeStepsInfo />", () => {
    test("Create time step info elements", () => {
        matches(<Fixture />);
    })
})