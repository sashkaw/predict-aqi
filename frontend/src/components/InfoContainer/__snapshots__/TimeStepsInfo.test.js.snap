// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeStepsInfo /> Create time step info elements 1`] = `
<div
  className="timesteps-container"
>
  <ul
    className="timesteps-list"
  >
    <li
      className="active"
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        11am
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        2
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        12pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        3
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        1pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        4
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        2pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        5
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        3pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        6
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        4pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        7
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        5pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        8
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        6pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        9
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        7pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        10
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        8pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        11
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        9pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        12
      </span>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <span
        className="timestep-name"
      >
        10pm
      </span>
      <span
        className="timestep-aqi"
      >
        AQI: 
        11
      </span>
    </li>
    <div
      className="clear"
    />
  </ul>
</div>
`;
